### ISSUE TITLE
backlinks buffer doesn't support non-latin characters

### ISSUE THREAD

**Issue by JonatanSahar** [protesilaos/denote#150]
When displaying the context along with the link I get:

> 20230331T161033--      -            -                -      -            -      __finance.org
> 11:                                 ,                                                                                  ? [[denote:20230331T161601][                     ]]
> 

which is actually a bunch of escaped characters:
![image](https://user-images.githubusercontent.com/18332012/229131284-6eb0799a-deb6-44e7-9894-aaf2f9079588.png)

Any idea why that is?
Thanks!

**Comment by protesilaos** [comment]
Hello @JonatanSahar!

Can you share this text with me?  I will try to reproduce it on my end.

**Comment by JonatanSahar** [comment]
shure.
this is a test note:

> #+title:      test
> #+date:       [2023-03-31 Fri 18:31]
> #+filetags:   :finance:
> #+identifier: 20230331T183136
> 

and this is a link to it with text:

> טקסט לבדיקה של קישורים ל [[denote:20230331T183136][test]]

**Comment by protesilaos** [comment]
Just to clarify that I also tested it with:

```elisp
(setq denote-backlinks-show-context t)
```

**Comment by JonatanSahar** [comment]
Thanks for looking into it.
with context set to `t` did you get the hebrew text that was on the same line with the link?
did you put the hebrew text I posted (טקסט לבדיקה של קישורים ל) with the link?

**Comment by protesilaos** [comment]
Just to be sure that we are testing the same thing, can you prepare 2 or 3 demo files in Hebrew and send them to me via email?  My email is: <public@protesilaos.com>

**Comment by JonatanSahar** [comment]
sure thing.

**Comment by JonatanSahar** [comment]
I just called `find-file` on my notes dir, and it shows file names correctly, however - calling `projectile-find-file` renders hebrew like the image in above. So this is not a denote issue.
Do you have an idea how I might isolate the problem though? it looks like different kinds of buffers support different encodings?

**Comment by protesilaos** [comment]
I do not know what is happening here.  Maybe some other Doom user can provide insights?

For the backlinks buffer in particular, I still get the expected results with the files you (@JonatanSahar) sent me, with and without `(setq denote-backlinks-show-context t)`.

![2023-04-12_06:40:24_965x1056](https://user-images.githubusercontent.com/12828033/231343953-cff4cb17-f08c-4359-b58a-1d32a14a8a09.png)

![2023-04-12_06:40:50_965x1056](https://user-images.githubusercontent.com/12828033/231343955-45f4fcea-6986-4fae-94d2-0d8730571ce9.png)

**Comment by JonatanSahar** [comment]
I'll ask around some more on Doom discourse/d, although they've not been responsive there recently. 

A couple more points: the xref mechanism does work with hebrew when used through `consult-grep` and then `embark export`, and also to note that the title of the backlink buffer is displayed correctly for me ("Backlinks to..." the name if the current note), and only then xrefs are without the proper encoding. 

I tried to look at how those are produced, but the code is too complicated for my knowledge with elsip..

**Comment by JonatanSahar** [comment]
I;ve found a workaround!
if called through `universal-coding-system-argument` I can force the buffer to use `utf-8`.
Do you know if there's a way to make that change permanent somehow? calling `denote-link-backlinks`  with `utf-8` programatically?

**Comment by rLry** [comment]
> I've found a workaround! if called through `universal-coding-system-argument` I can force the buffer to use `utf-8`. Do you know if there's a way to make that change permanent somehow? calling `denote-link-backlinks` with `utf-8` programatically?

My note's name contains Chinese, and the same problem occurs.
I can't force buffer to use utf-8 even through `universal-coding-system-argument`

**Comment by protesilaos** [comment]
@rLry Just for me to understand this better: what operating system are you using?

**Comment by rLry** [comment]
> @rLry Just for me to understand this better: what operating system are you using?

win11 22h2
（ I have set the system default encoding to UTF-8.I tried changing it, but it didn't work.）

**Comment by protesilaos** [comment]
Maybe this is due to a shell environment variable that affects `grep` (specifically, the `xref-search-program`).  What does this give you in Emacs?

```elisp
(getenv "LANG")
```

**Comment by rLry** [comment]
> Maybe this is due to a shell environment variable that affects `grep` (specifically, the `xref-search-program`). What does this give you in Emacs?
> 
> ```emacs-lisp
> (getenv "LANG")
> ```

`CHS`

**Comment by protesilaos** [comment]
Do you use regular `M-x grep`?  If yes, can you try to search those files you have that contain Chinese characters in the file name?  Denote is not involved in these operations, but having a way to reproduce the problem means that we can report it on the Emacs bug tracker (and I am happy to help with that).

**Comment by rLry** [comment]
> Do you use regular `M-x grep`? If yes, can you try to search those files you have that contain Chinese characters in the file name? Denote is not involved in these operations, but having a way to reproduce the problem means that we can report it on the Emacs bug tracker (and I am happy to help with that).

Thanks. I'm now sure this issue is related to ripgrep searches. I will continue to try to solve it, maybe this is related to the confusing encoding of windows

**Comment by protesilaos** [comment]
@rLry Yes, this looks like an issue with Windows and the environment those command-line tools are running in.  Unfortunately, I am not familiar with Windows to be helpful here.

**Comment by rLry** [comment]
> @rLry Yes, this looks like an issue with Windows and the environment those command-line tools are running in. Unfortunately, I am not familiar with Windows to be helpful here.

I think I found the solution in Windows, although it may be a little late.
1. `scoop install busybox` to ensure the `xargs` installed.
Alternatively to use xargs of git as mentioned in the manual.
2. Set the correct encoding according to the text  used. For example, I am a Chinese user,
> (setq-default process-coding-system-alist
                     '(("[pP][lL][iI][nN][kK]" utf-8-dos . gbk-dos)
                       ("[cC][mM][dD][pP][rR][oO][xX][yY]" utf-8-dos . gbk-dos)
                       ("[rR][gG]" utf-8-dos . gbk-dos)
                       ("rg" utf-8 . gbk-dos)
                       ("ug" utf-8 . gbk-dos)
                       ("ugrep" utf-8 . gbk-dos)
                       ("explorer" gbk-dos . gbk-dos)
                       ("*" utf-8 . utf-8)))

My knowledge of emacs is very poor, if someone has better code that would be great.

**Comment by protesilaos** [comment]
@rLry So `xargs` is a prerequisite here? Is that coming from `busybox` only or is there some GNU package as well.

What you do with `process-coding-system-alist` seems reasonable. If anyone can test this, please note it here. We will expand the manual accordingly.

**Comment by rLry** [comment]
> @rLry So `xargs` is a prerequisite here? Is that coming from `busybox` only or is there some GNU package as well.
> 
> What you do with `process-coding-system-alist` seems reasonable. If anyone can test this, please note it here. We will expand the manual accordingly.

yes, it is prerequisite.
I installed busybox specifically because xref in git wasn't working on my computer.

**Comment by TomoeMami** [comment]
> yes, it is prerequisite. I installed busybox specifically because xref in git wasn't working on my computer.

Based on my current testing, only `xargs` in `busybox` works properly. Neither `gow`’s nor `git`’s `xargs` implementations function correctly - they simply display “No backlinks for query”. (I’m also a Chinese user, with a Chinese Windows 11 24H2, using `scoop` for `busybox` `gow` `git`.)

**Comment by protesilaos** [comment]
Hello @TomoeMami! I am afraid this is a problem with the underlying Xref functionality that is built into Emacs. There is nothing we can do on the Denote side. I am nevertheless re-opening this issue in case there is some other setup that works on Windows.
